lib/primitives/block.js

```diff
@@ -839,6 +841,46 @@ class Block extends AbstractBlock {
   static isBlock(obj) {
     return obj instanceof Block;
   }
+
+  /*
+   * Get block filter (BIP 158)
+   * @param {CoinView} view
+   * @returns {Object} See {@link Golomb}
+   */
+
+  toFilter(view) {
```

Comment:
There can be different types of filters, so I think a single method `toFilter` is short sighted. The filter that is built inside of this method is defined as the `Basic` filter, defined as type `0x00`. Would it be possible to add a `type` parameter to this method or some other form of refactoring that allows for easily playing with new filter types?

I'd personally like to create a new filter type that includes all of the outpoints consumed in a block, so that you can easily detect when a UTXO is spent.

See https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#block-filters

Reply:
Maybe there is a wrapper `toBasicFilter` which calls `toFilter(filterTypes.BASIC, view)`. This could result in `toFilter` growing to be a massive function, as the logic would be al different for each filter type.

Maybe we are just explicit and rename this method `toBasicFilter` so that we can add other filter creation methods easily.

There is also the question in the REST API/RPC of being able to query additional types, maybe we allow either the type int or the type value by keeping a map

```
filterTypesByValue = {
  'BASIC': 0x00
}
filterTypes = {
  0x00: 'BASIC'
}
```

Then a user can send a request for:

`GET /filter/<hash|height>?type=basic`

or

`$ bcoin-cli rpc getblockfilter <hash> <type>`

Type `basic` could be the default if no other method is specified.

Reply:
I don't think this change from `toFilter` to `toBasicFilter` is necessary. There is consistence use of language for block filter. If there are additional filters, I think it would then be necessary to articulate any difference in the method and elsewhere. To abstract it at this point is too soon.

Reply:
I'd like to be able to create an additional filter type, one that we could eventually try to get into bip 158. From above:

> I'd personally like to create a new filter type that includes all of the outpoints consumed in a block, so that you can easily detect when a UTXO is spent.

Do you think there will never be additional filter types? I think that this is an opportunity for bcoin to be ahead on something in the ecosystem. To not abstract it at this point will make it more difficult to be innovative.

Reply:
Yeah BIP 157/158 describes multiple filter types, but they dropped the extended filter, all the steps related to multiple filter types seem like an exercise in futility. As for custom filter types, most of the code is already fairly modular imo so like @braydonf said it's probably too early to abstract atm.

Reply:
The indexes, especially this one, is very quick to regenerate. If there are any necessary changes that come up in the future, there shouldn't be an issue with making any necessary updates.

Reply:
There are several unanswered questions that would come up in the process of adding another filter. As an additional type is currently undefined, the use case is also undefined, and has currently unknown requirements. It's certainly an area that can be researched for utility.